### PR TITLE
fix(serializer): put replaces embed collection

### LINKED ISSUE #5587: 3.1.11 - PUT does not work with ManyToMany deep nested relations
**API Platform version(s) affected**: `3.1.11`

**Description**  
`PATCH` does not work with ManyToMany deep nested relations.
But now neither does PUT `standard_put: true` in `3.1.11`

**Possible Solution**  
Revert latest changes made in `ItemNormalizer.php`

### PR DESCRIPTION
**Pull request by soyuka**
fixes #5587 #5603

### CODE REVIEW

**Comment by mdieudonne** [comment]
This works for me! Thanks.

**Comment by jelovac** [comment]
Hi, when can we expect a release for this fix?

**Comment by jonag** [comment]
Hello @soyuka, I have updated API Platform to the last version (v3.1.13) and I think that this PR introduced a BC Break.

In order to get arround [this issue](https://github.com/api-platform/core/issues/5559) I wrote a custom Normalizer to unset the `OBJECT_TO_POPULATE` and thus allowing me to update a collection operation while PATCHing a resource.

<details>
  <summary>Here is my custom Normalizer:</summary>

  ```php
  <?php
  
  namespace App\Normalizer;
  
  use ApiPlatform\Metadata\Patch;
  use Doctrine\Common\Collections\Collection;
  use Symfony\Component\DependencyInjection\Attribute\AsDecorator;
  use Symfony\Component\DependencyInjection\Attribute\AutowireDecorated;
  use Symfony\Component\Serializer\Normalizer\AbstractNormalizer;
  use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
  use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
  use Symfony\Component\Serializer\SerializerAwareInterface;
  use Symfony\Component\Serializer\SerializerInterface;
  
  #[AsDecorator('api_platform.jsonld.normalizer.item')]
  final readonly class AllowPatchEmbeddedCollectionRelationNormalizer implements NormalizerInterface, DenormalizerInterface, SerializerAwareInterface
  {
      public function __construct(
          #[AutowireDecorated] private NormalizerInterface&DenormalizerInterface $decorated,
      ) {
      }
  
      public function supportsNormalization($data, $format = null, array $context = []): bool
      {
          return $this->decorated->supportsNormalization($data, $format);
      }
  
      /**
       * {@inheritdoc}
       */
      public function normalize($object, $format = null, array $context = []): float|array|\ArrayObject|bool|int|string|null
      {
          return $this->decorated->normalize($object, $format, $context);
      }
  
      /**
       * {@inheritdoc}
       */
      public function denormalize($data, $type, $format = null, array $context = []): mixed
      {
          if (
              ($context['operation'] ?? null) instanceof Patch
              && ($context[AbstractNormalizer::OBJECT_TO_POPULATE] ?? null) instanceof Collection
          ) {
              unset($context[AbstractNormalizer::OBJECT_TO_POPULATE]);
          }
  
          return $this->decorated->denormalize($data, $type, $format, $context);
      }
  
      /**
       * {@inheritdoc}
       */
      public function supportsDenormalization($data, $type, $format = null, array $context = []): bool
      {
          return $this->decorated->supportsDenormalization($data, $type, $format);
      }
  
      /**
       * {@inheritdoc}
       */
      public function setSerializer(SerializerInterface $serializer): void
      {
          if ($this->decorated instanceof SerializerAwareInterface) {
              $this->decorated->setSerializer($serializer);
          }
      }
  }
  ```
</details>

However this PR replaces the original operation with a `Get` operation and therefore breaks my workaround.

Please tell me if I should rather open a new issue to discuss it (I can also upload a repo with my failing test if needed).

**Comment by Aerendir** [comment]
@soyuka , I confirm this change is a BC change: it broke my tests where:

- PUT Resource with POST SubResource
- PUT Resource with PUT SubResource
- PUT Resource with DELETE SubResource

More, it may also be wrong: the problem reported by @mdieudonne seems to relate to the name of placeholders: all his placeholders are named "id": this causes a conflict.

Removing the child context removes the conflict, but also breaks my PR #5546 : the child context is required to get the subresources.

So, maybe that simply changing the name of the placeholders, avoiding using the same name for different placeholders, would have resolved the problem reported by @mdieudonne .

Anyway, this PR makes impossible to use, for resources IDs, placeholders that have a different name than "id" (that is exactly what I fixed with my PR - and that was something known as there was a todo comment that remembered to fix it).

WDYT?

**Comment by soyuka** [comment]
I noted both of your comments they don't seem to have the same concerns... I'll handle these asap, please stay on the previous 3.1.12 for now if you can.

File: src/Serializer/AbstractItemNormalizer.php [hunk 1]
```diff
@@ -773,7 +778,6 @@ private function createAndValidateAttributeValue(string $attribute, mixed $value
             && $this->resourceClassResolver->isResourceClass($className)
         ) {
             $resourceClass = $this->resourceClassResolver->getResourceClass(null, $className);
-            $context['resource_class'] = $resourceClass;
```
Comment (soyuka) [inline src/Serializer/AbstractItemNormalizer.php:776]: this looks like this should be reverter

**Comment by soyuka** [comment]
@jonag can you use `root_operation` ?

**Comment by Aerendir** [comment]
> I noted both of your comments they don't seem to have the same concerns... I'll handle these asap, please stay on the previous 3.1.12 for now if you can.

Yes, they are two different problems...

**Comment by jonag** [comment]
> @jonag can you use `root_operation` ?

I should find this key in the context passed to the normalizer? Because this key is currently not set when I dump the context on my custom normalizer.

**Comment by soyuka** [comment]
https://github.com/api-platform/core/blob/92a81f024541054b9322e7457b75c721261e14e0/src/Hydra/Serializer/CollectionNormalizer.php#L92 it should be there when using Hydra, I may add this to the SerializerContextBuilder but I'd like for you to try it first if it's possible?

**Comment by nxtpge** [comment]
Hello @jonag and @soyuka, is the fix still ongoing?

**Comment by jonag** [comment]
Hello @nxtpge, if you are talking about the property `root_operation`, I forgot to answer to @soyuka but it's indeed already available in the context passed to the normalizer.

**Pull request merged**